torch/csrc/jit/codegen/cuda/transform_view.cpp

```diff
@@ -0,0 +1,678 @@
+    for (auto s : new_view_) {
+      output << " " << s;
+    }
+    output << std::endl;
+
+    output << "===============================" << std::endl;
+    for (auto move : trivial_reduction_transforms_) {
+      move->toString(output);
+    }
+    for (auto move : view_transforms_) {
+      move->toString(output);
+    }
+    for (auto move : broadcast_transforms_) {
+      move->toString(output);
+    }
+    output << "===============================" << std::endl;
+    return output.str();
+  }
+
+  void addMergeTransform(bool lhs, bool rfactor) {
+    if (lhs) {
+      view_transforms_.push_back(new MergeTransform(state_, rfactor));
+    } else {
+      // TODO: original_view_index points to the merge rhs
+      // The invariant for merge transform is index = merge_rhs_position-1
+      ViewIndexState clone_state(state_);
+      --clone_state.original_view_index;
+      view_transforms_.push_back(new MergeTransform(clone_state, rfactor));
+    }
+    ++state_.merge_offset;
+  }
+
+  void addSplitTransform(size_t split_factor, bool rfactor) {
+    view_transforms_.push_back(
+        new SplitTransform(state_, rfactor, split_factor));
+    ++state_.split_offset;
+    ++state_.new_view_index;
+  }
+
+  void addKeepTransform(bool rfactor) {
+    if (!rfactor) {
+      view_transforms_.push_back(new KeepTransform(state_));
+    }
+    ++state_.new_view_index;
+    ++state_.original_view_index;
+    ++state_.transform_view_index;
+  }
+
+  void addBroadcastTransform() {
+    broadcast_transforms_.push_back(new BroadcastTransform(state_));
+    ++state_.broadcast_offset;
+    ++state_.new_view_index;
+  }
+
+  void addTrivialReductionTransform() {
+    trivial_reduction_transforms_.push_back(
+        new TrivialReductionTransform(state_));
+    ++state_.trivial_reduction_offset;
+  }
+
+  bool validate() const {
+    if (state_.new_view_index != new_view_.size() ||
+        state_.original_view_index != original_view_.size() ||
+        state_.transform_view_index != transform_view_.size()) {
+      return false;
+    }
+    return true;
+  }
+
+  //! Utility class for merging multiple axes together.
+  class AnalyzeMergeInterface {
+   protected:
+    void handle(bool rfactor, bool lhs) {
+      getMergeOffset();
+
+      bool any_merge = false;
+      for (size_t idx = 0; idx < offset_; ++idx) {
+        if (avt_->root_domain_[state_.original_view_index]
+                ->isImplicitBroadcast()) {
+          avt_->addTrivialReductionTransform();
+        } else {
+          avt_->addMergeTransform(lhs, rfactor || any_merge);
+          any_merge = true;
+        }
+        stateUpdate();
+      }
+
+      epilogue(rfactor || any_merge);
+    }
+
+    AnalyzeMergeInterface(
+        AnalyzeViewTransformation* avt,
+        ViewIndexState& state,
+        size_t current_size = 1)
+        : avt_(avt), state_(state), current_size_(current_size) {}
+
+    virtual size_t getCurrentIndex() const = 0;
+    virtual bool automaticFail() const = 0;
```

Comment:
The naming doesn't seem to represent what it is supposed to do (if I understand what it actually is). Maybe something like `shouldContinue`?